File: contests/lc/2043.SimpleBankSystem/solution.go
package _043_SimpleBankSystem

// Bank simulates a set of bank accounts with basic operations.
type Bank struct {
	balance []int64
}

// Constructor creates a Bank with a defensive copy of the initial balances.
func Constructor(balance []int64) Bank {
	balances := make([]int64, len(balance))
	copy(balances, balance)
	return Bank{balance: balances}
}

// Transfer moves funds from account1 to account2 when both accounts exist and have sufficient funds.
func (b *Bank) Transfer(account1 int, account2 int, money int64) bool {
	if !b.validAccount(account1) || !b.validAccount(account2) {
		return false
	}
	idx1 := account1 - 1
	idx2 := account2 - 1
	if b.balance[idx1] < money {
		return false
	}
	b.balance[idx1] -= money
	b.balance[idx2] += money
	return true
}

// Deposit adds funds to the given account when it exists.
func (b *Bank) Deposit(account int, money int64) bool {
	if !b.validAccount(account) {
		return false
	}
	b.balance[account-1] += money
	return true
}

// Withdraw removes funds from the given account if it exists and has enough balance.
func (b *Bank) Withdraw(account int, money int64) bool {
	if !b.validAccount(account) {
		return false
	}
	idx := account - 1
	if b.balance[idx] < money {
		return false
	}
	b.balance[idx] -= money
	return true
}

func (b *Bank) validAccount(account int) bool {
	return account >= 1 && account <= len(b.balance)
}

File: contests/lc/2043.SimpleBankSystem/solution.py
from __future__ import annotations

from typing import List


class Bank:
    """Simulates bank accounts with transfer, deposit, and withdraw operations."""

    def __init__(self, balance: List[int]) -> None:
        # Copy to prevent callers from mutating the internal state.
        self._balance: List[int] = list(balance)

    def transfer(self, account1: int, account2: int, money: int) -> bool:
        if not self._valid_account(account1) or not self._valid_account(account2):
            return False
        idx1 = account1 - 1
        idx2 = account2 - 1
        if self._balance[idx1] < money:
            return False
        self._balance[idx1] -= money
        self._balance[idx2] += money
        return True

    def deposit(self, account: int, money: int) -> bool:
        if not self._valid_account(account):
            return False
        self._balance[account - 1] += money
        return True

    def withdraw(self, account: int, money: int) -> bool:
        if not self._valid_account(account):
            return False
        idx = account - 1
        if self._balance[idx] < money:
            return False
        self._balance[idx] -= money
        return True

    def _valid_account(self, account: int) -> bool:
        return 1 <= account <= len(self._balance)

    @property
    def balance(self) -> List[int]:
        return list(self._balance)
